In [1]:
#Importing libraries as pd

import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

In [2]:
#creating path
path = r'C:\Users\rbrand\12-28-2024 Instacart Basket Analysis'

In [3]:
#3.Importing customers.csv data set from original data.
df_cust = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'customers.csv'), index_col = False)

In [4]:
df_cust.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

In [5]:
#4.Renaming Surnam column
df_cust.rename(columns={'Surnam': 'Last Name'}, inplace=True)

In [6]:
print(df_cust.columns)

Index(['user_id', 'First Name', 'Last Name', 'Gender', 'STATE', 'Age',
       'date_joined', 'n_dependants', 'fam_status', 'income'],
      dtype='object')


In [7]:
# Droping one more column that don't add anything to analysis. I did remove two more, fam status and n_dependants bfore.
columns_to_drop = ['date_joined','n_dependants']
df_cust.drop(columns=columns_to_drop, inplace=True)

In [8]:
columns_to_drop = ['fam_status']
df_cust.drop(columns=columns_to_drop, inplace=True)

In [9]:
print(df_cust.columns)

Index(['user_id', 'First Name', 'Last Name', 'Gender', 'STATE', 'Age',
       'income'],
      dtype='object')


In [10]:
#5. Checking for missing values
print(df_cust.isnull().sum())

user_id           0
First Name    11259
Last Name         0
Gender            0
STATE             0
Age               0
income            0
dtype: int64


In [11]:
#Checkinc duplicates
print(df_cust.duplicated().sum())

0


In [12]:
#Checking data types
print(df_cust.dtypes)

user_id        int64
First Name    object
Last Name     object
Gender        object
STATE         object
Age            int64
income         int64
dtype: object


For missing values, the column "First Name" has 11,259 missing values.The column is not crucial for your analysis so we can proceed without it. I will simply drop the rows with missing values.

here are no duplicate rows in your dataset, which is good.

The data types seem appropriate for each column.

In [13]:
# Dropping "First Name" column
columns_to_drop = ['First Name']
df_cust.drop(columns=columns_to_drop, inplace=True)

In [14]:
print(df_cust.columns)

Index(['user_id', 'Last Name', 'Gender', 'STATE', 'Age', 'income'], dtype='object')


In [15]:
#6.Importing ords_prods_merge.csv data set from prepared data to combine both
file_path = os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merge.pkl')

In [16]:
#Using the read.pickle to import data frame.
ords_prods_merge = pd.read_pickle(file_path)

In [17]:
#Checking the data frame import.
ords_prods_merge.head()

Unnamed: 0_x  order_id  user_id  order_number  orders_day_of_week  \
0             0   2539329        1             1                   2   
1             1   2398795        1             2                   3   
2             2    473747        1             3                   3   
3             3   2254736        1             4                   4   
4             4    431534        1             5                   4   

   order_hour_of_day  order_frequency_days  product_id  add_to_cart_order  \
0                  8                   NaN         196                  1   
1                  7                  15.0         196                  1   
2                 12                  21.0         196                  1   
3                  7                  29.0         196                  1   
4                 15                  28.0         196                  1   

   reordered  ... prices  merge_flag    price_range_loc     busiest_day  \
0          0  ...    9.0        both  Mid-range product  Regularly busy   
1          1  ...    9.0        both  Mid-range product  Regularly busy   
2          1  ...    9.0        both  Mid-range product  Regularly busy   
3          1  ...    9.0        both  Mid-range product      Least busy   
4          1  ...    9.0        both  Mid-range product      Least busy   

   new_busiest_days  busiest_period_of_day max_order  loyalty_flag  \
0    Regularly busy         Average orders        10  New customer   
1      Slowest days         Average orders        10  New customer   
2      Slowest days            Most orders        10  New customer   
3      Slowest days         Average orders        10  New customer   
4      Slowest days            Most orders        10  New customer   

  spending_flag   order_frequency_flag  
0   Low spender  Non-frequent customer  
1   Low spender  Non-frequent customer  
2   Low spender  Non-frequent customer  
3   Low spender  Non-frequent customer  
4   Low spender  Non-frequent customer  

[5 rows x 25 columns]

In [21]:
#Exporting df_cust to prepared data as customers.csv since in order to solve memory errors.
df_cust.to_csv(os.path.join(path, '02 Data','Prepared Data', 'customers.csv'))

In [ ]:
--------------------------------------------------------------------------------------------

In [1]:
#Importing libraries as pd

import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

In [2]:
#creating path
path = r'C:\Users\rbrand\12-28-2024 Instacart Basket Analysis'

In [4]:
#3.Importing customers.csv data set from prepared data.
df_cust = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'customers.csv'), index_col = False)

In [5]:
df_cust.head()

Unnamed: 0  user_id Last Name  Gender       STATE  Age  income
0           0    26711  Esquivel  Female    Missouri   48  165665
1           1    33890      Hart  Female  New Mexico   36   59285
2           2    65803    Farley    Male       Idaho   35   99568
3           3   125935     Hicks  Female        Iowa   40   42049
4           4   130797   Gilmore  Female    Maryland   26   40374

In [8]:
# Define the file path for ords_prods_merge.pkl
file_path = os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merge.pkl')

In [9]:
# Load the ords_prods_merge dataframe from the pickle file
ords_prods_merge = pd.read_pickle(file_path)

In [15]:
# Reducing ords_prods_merge dataframe to 1,000,000 rows due memory erros
ords_prods_merge = ords_prods_merge.head(1000000)

In [16]:
print("ords_prods_merge rows:", len(ords_prods_merge))

ords_prods_merge rows: 1000000


In [17]:
#Using 'user_id' is the key column for merging
combined_df = pd.merge(ords_prods_merge, df_cust, on='user_id', how='left')

In [21]:
combined_df.head()

Unnamed: 0_x  order_id  user_id  order_number  orders_day_of_week  \
0             0   2539329        1             1                   2   
1             1   2398795        1             2                   3   
2             2    473747        1             3                   3   
3             3   2254736        1             4                   4   
4             4    431534        1             5                   4   

   order_hour_of_day  order_frequency_days  product_id  add_to_cart_order  \
0                  8                   NaN         196                  1   
1                  7                  15.0         196                  1   
2                 12                  21.0         196                  1   
3                  7                  29.0         196                  1   
4                 15                  28.0         196                  1   

   reordered  ... max_order  loyalty_flag spending_flag  \
0          0  ...        10  New customer   Low spender   
1          1  ...        10  New customer   Low spender   
2          1  ...        10  New customer   Low spender   
3          1  ...        10  New customer   Low spender   
4          1  ...        10  New customer   Low spender   

    order_frequency_flag  Unnamed: 0  Last Name  Gender    STATE Age income  
0  Non-frequent customer      134862     Nguyen  Female  Alabama  31  40423  
1  Non-frequent customer      134862     Nguyen  Female  Alabama  31  40423  
2  Non-frequent customer      134862     Nguyen  Female  Alabama  31  40423  
3  Non-frequent customer      134862     Nguyen  Female  Alabama  31  40423  
4  Non-frequent customer      134862     Nguyen  Female  Alabama  31  40423  

[5 rows x 31 columns]

In [22]:
# Checking for missing values
missing_values = combined_df.isnull().sum()
print("Missing Values:\n", missing_values)

Missing Values:
 Unnamed: 0_x                 0
order_id                     0
user_id                      0
order_number                 0
orders_day_of_week           0
order_hour_of_day            0
order_frequency_days     60085
product_id                   0
add_to_cart_order            0
reordered                    0
_merge                       0
Unnamed: 0_y                 0
product_name                 0
aisle_id                     0
department_id                0
prices                       0
merge_flag                   0
price_range_loc              0
busiest_day                  0
new_busiest_days             0
busiest_period_of_day        0
max_order                    0
loyalty_flag                 0
spending_flag                0
order_frequency_flag         0
Unnamed: 0                   0
Last Name                    0
Gender                       0
STATE                        0
Age                          0
income                       0
dtype: int64


In [23]:
# Checking for duplicate rows
duplicates = combined_df.duplicated().sum()
print("Number of Duplicate Rows:", duplicates)

Number of Duplicate Rows: 0


In [24]:
# Droping rows with missing values in 'order_frequency_days' for data accurancy
combined_df = combined_df.dropna(subset=['order_frequency_days'])

In [25]:
#Double checking if missing values are resolved
missing_values_after_drop = combined_df.isnull().sum()
print("Missing Values After Drop:\n", missing_values_after_drop)

Missing Values After Drop:
 Unnamed: 0_x             0
order_id                 0
user_id                  0
order_number             0
orders_day_of_week       0
order_hour_of_day        0
order_frequency_days     0
product_id               0
add_to_cart_order        0
reordered                0
_merge                   0
Unnamed: 0_y             0
product_name             0
aisle_id                 0
department_id            0
prices                   0
merge_flag               0
price_range_loc          0
busiest_day              0
new_busiest_days         0
busiest_period_of_day    0
max_order                0
loyalty_flag             0
spending_flag            0
order_frequency_flag     0
Unnamed: 0               0
Last Name                0
Gender                   0
STATE                    0
Age                      0
income                   0
dtype: int64


In [26]:
combined_df.head()

Unnamed: 0_x  order_id  user_id  order_number  orders_day_of_week  \
1             1   2398795        1             2                   3   
2             2    473747        1             3                   3   
3             3   2254736        1             4                   4   
4             4    431534        1             5                   4   
5             5   3367565        1             6                   2   

   order_hour_of_day  order_frequency_days  product_id  add_to_cart_order  \
1                  7                  15.0         196                  1   
2                 12                  21.0         196                  1   
3                  7                  29.0         196                  1   
4                 15                  28.0         196                  1   
5                  7                  19.0         196                  1   

   reordered  ... max_order  loyalty_flag spending_flag  \
1          1  ...        10  New customer   Low spender   
2          1  ...        10  New customer   Low spender   
3          1  ...        10  New customer   Low spender   
4          1  ...        10  New customer   Low spender   
5          1  ...        10  New customer   Low spender   

    order_frequency_flag  Unnamed: 0  Last Name  Gender    STATE Age income  
1  Non-frequent customer      134862     Nguyen  Female  Alabama  31  40423  
2  Non-frequent customer      134862     Nguyen  Female  Alabama  31  40423  
3  Non-frequent customer      134862     Nguyen  Female  Alabama  31  40423  
4  Non-frequent customer      134862     Nguyen  Female  Alabama  31  40423  
5  Non-frequent customer      134862     Nguyen  Female  Alabama  31  40423  

[5 rows x 31 columns]

In [28]:
# Droping one more column that don't add anything to analysis. I did remove two more, fam status and n_dependants bfore.
columns_to_drop = ['Unnamed: 0', 'Unnamed: 0_x']
combined_df.drop(columns=columns_to_drop, inplace=True)

In [30]:
combined_df.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   
5   3367565        1             6                   2                  7   

   order_frequency_days  product_id  add_to_cart_order  reordered _merge  ...  \
1                  15.0         196                  1          1   both  ...   
2                  21.0         196                  1          1   both  ...   
3                  29.0         196                  1          1   both  ...   
4                  28.0         196                  1          1   both  ...   
5                  19.0         196                  1          1   both  ...   

   busiest_period_of_day max_order  loyalty_flag  spending_flag  \
1         Average orders        10  New customer    Low spender   
2            Most orders        10  New customer    Low spender   
3         Average orders        10  New customer    Low spender   
4            Most orders        10  New customer    Low spender   
5         Average orders        10  New customer    Low spender   

    order_frequency_flag Last Name  Gender    STATE Age income  
1  Non-frequent customer    Nguyen  Female  Alabama  31  40423  
2  Non-frequent customer    Nguyen  Female  Alabama  31  40423  
3  Non-frequent customer    Nguyen  Female  Alabama  31  40423  
4  Non-frequent customer    Nguyen  Female  Alabama  31  40423  
5  Non-frequent customer    Nguyen  Female  Alabama  31  40423  

[5 rows x 29 columns]

In [31]:
#Exporting the combined_df DataFrame as a Pickle file.
combined_df.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'combined_df.pkl'))